compiler/p/codegen/OMRLoadStoreHandler.cpp

```diff
@@ -0,0 +1,393 @@
+
+            generateTrg1MemInstruction(cg, TR::InstOpCode::lwz, node, trgReg->getHighOrder(), spillRefHi);
+            generateTrg1MemInstruction(cg, TR::InstOpCode::lwz, node, trgReg->getLowOrder(), spillRefLo);
+
+            cg->freeSpill(spillLocation, 8, 0);
+            cg->stopUsingRegister(tempReg);
+            }
+        }
+    else
+#endif
+        {
+        TR::MemoryReference *memRefHi = TR::MemoryReference::createWithMemRef(cg, node, *memRef, 0, 4);
+        TR::MemoryReference *memRefLo = TR::MemoryReference::createWithMemRef(cg, node, *memRef, 4, 4);
+
+        generateTrg1MemInstruction(cg, TR::InstOpCode::lwz, node, trgReg->getHighOrder(), memRefHi);
+        generateTrg1MemInstruction(cg, TR::InstOpCode::lwz, node, trgReg->getLowOrder(), memRefLo);
+        }
+    }
+
+#ifdef J9_PROJECT_SPECIFIC
+enum class StoreSyncRequirements
+    {
+    None,
+    AtomicWrite,
+    LazySet,
+    Full
+    };
+
+StoreSyncRequirements getStoreSyncRequirements(TR::CodeGenerator *cg, TR::Node *node)
+    {
+    if (node->getSymbol()->isSyncVolatile())
+        {
+        // Since non-volatiles are implemented as two separate stores, we must use a special sequence to perform the
+        // store in a single instruction even when SMP is disabled.
+        if (!cg->comp()->target().isSMP())
+            return StoreSyncRequirements::AtomicWrite;
+
+        TR_OpaqueMethodBlock *caller = node->getOwningMethod();
+
+        if (caller && !cg->comp()->compileRelocatableCode())
+            {
+            TR_ResolvedMethod *m = cg->comp()->fe()->createResolvedMethod(cg->trMemory(), caller, node->getSymbolReference()->getOwningMethod(cg->comp()));
+            if (m->getRecognizedMethod() == TR::java_util_concurrent_atomic_AtomicInteger_lazySet)
+                return StoreSyncRequirements::LazySet;
+            }
+
+        return StoreSyncRequirements::Full;
+        }
+    else if (node->getSymbol()->isShadow() && node->getSymbol()->isOrdered() && cg->comp()->target().isSMP())
+        {
+        return StoreSyncRequirements::LazySet;
+        }
+
+    return StoreSyncRequirements::None;
+    }
+#endif
+
+void OMR::Power::LoadStoreHandlerImpl::generateStoreSequence(TR::CodeGenerator *cg, TR::Register *srcReg, TR::Node *node, TR::MemoryReference *memRef, TR::InstOpCode::Mnemonic storeOp)
+    {
+#ifdef J9_PROJECT_SPECIFIC
+    StoreSyncRequirements sync = getStoreSyncRequirements(cg, node);
+
+    if (sync == StoreSyncRequirements::LazySet || sync == StoreSyncRequirements::Full)
+        generateInstruction(cg, TR::InstOpCode::lwsync, node);
+#endif
+
+    generateMemSrc1Instruction(cg, storeOp, node, memRef, srcReg);
+
+#ifdef J9_PROJECT_SPECIFIC
+    if (sync == StoreSyncRequirements::Full)
+        TR::TreeEvaluator::postSyncConditions(node, cg, srcReg, memRef, TR::InstOpCode::sync);
+#endif
+    }
```

Comment:
I am assuming the unresolved load/store layout is not changed as this takes effect, since the memory barriers are emitted proactively to occupy the space.  they will be patched with nop  if the resolve proves to be none volatile.  the patching is based on the layout, so it is critical for correctness.

Reply:
This change shouldn't cause any unresolved stores that previously had synchronization barriers to lose them, but may cause unresolved stores that would previously be lacking these barriers to gain them for correctness. Theoretically, all unresolved stores _should_ have a synchronization requirement of `Full` due to being marked as volatile (though there was at least one instance where an unresolved symref was somehow not marked as volatile IIRC).

Regardless, missing synchronization barriers cannot cause unresolved handling to patch incorrectly to my knowledge. To handle the patching, `UnresolvedDataSnippet` actually has a flag that indicates whether there are barriers that should be patched out if the resolved symref is non-volatile. This flag is only set if `postSyncConditions` is called, so the unresolved handling code should never try to patch out a non-existent `sync` instruction. This is coincidentally the reason that the previously missing synchronization barriers never caused issues with patching.

Reply:
not only the sync(s), but the instructions in between as well.  the complete layout is critical to the correctness.  this surely will be exercised when rtResolve,gcOnResolve tests are launched.

Reply:
The layout has not been changed. If this is surely exercised by the `rtResolve,gcOnResolve` tests, then this should definitely be correct as I have run this against as many OpenJ9 test suites as I could.